match2/commons/match_group_display_bracket.lua

```diff
@@ -1,738 +1,745 @@
 
-    -- parse bracketdata
-    self.referencedIds = {}
-    self.type = bracketdata.type or 'error'
-    self.header = utils.misc.emptyOr(bracketdata.header or nil)
-    self.skipround = bracketdata.skipround or 0
-    if (self.skipround == 'true') then
-        self.skipround = 1
-    else
-        self.skipround = tonumber(self.skipround) or 0
-    end
-    self.qualskip = bracketdata.qualskip or 0
-    if (self.qualskip == 'true') then
-        self.qualskip = 1
-    else
-        self.qualskip = tonumber(self.qualskip) or 0
-    end
-    self.qualwin = utils.misc.readBool(bracketdata.qualwin or 0)
-    self.quallose = utils.misc.readBool(bracketdata.quallose or 0)
-    self.upper = utils.misc.emptyOr(bracketdata.toupper or nil)
-    self.lower = utils.misc.emptyOr(bracketdata.tolower or nil)
-    self.qualwinLiteral = utils.misc.emptyOr(bracketdata.qualwinLiteral or '')
-    self.qualloseLiteral = utils.misc.emptyOr(bracketdata.qualloseLiteral or '')
-    self.bracketreset = utils.misc.emptyOr(bracketdata.bracketreset or nil)
-    self.thirdplace = utils.misc.emptyOr(bracketdata.thirdplace or nil)
-    self.opponent1Raw.displaytype = 'bracket'
-    self.opponent2Raw.displaytype = 'bracket'
-
-    -- set bracket specific visual parameters
-    self.minDisplayHeight = 0
-    if (self.thirdplace ~= nil) and has3rd then
-        self.minDisplayHeight = 3 * self.matchHeight + 3 * self.padding + self.headerHeight + self.headerHeight2
-    elseif (self.quallose) then
-        self.minDisplayHeight = 3 * self.matchHeight + self.headerHeight
+        -- Height of this round, including the header but excluding the 3rd place match and qualifier rounds.
+        local height = headerFullHeight 
+            + math.max(
+                lowerMarginTop + heightSums[#heightSums], 
+                matchMarginTop + matchHeight + config.matchMargin
+            )
+        
+        return {
+            height = height,
+            lowerMarginTop = lowerMarginTop,
+            matchHeight = matchHeight,
+            matchMarginTop = matchMarginTop,
+            mid = mid,
+            showHeader = showHeader,
+        }
+    end)
+
+    local layoutsByMatchId = {}
+    for matchId, _ in pairs(matchesById) do
+        layoutsByMatchId[matchId] = computeNodeLayout(matchId)
     end
 
-    -- store referenced ids
-    if (self.upper ~= nil) then
-        self.referencedIds[self.upper] = true
-    end
-    if (self.lower ~= nil) then
-        self.referencedIds[self.lower] = true
-    end
-    if (self.bracketreset ~= nil) then
-        self.referencedIds[self.bracketreset] = true
-    end
-    if (self.thirdplace ~= nil) then
-        self.referencedIds[self.thirdplace] = true
+    -- Matches without upper matches
+    local headMatchIds = {}
+    for matchId, _ in pairs(matchesById) do
+        if not upperMatchIds[matchId] 
+            and not LuaUtils.string.endsWith(matchId, 'RxMTP') 
+            and not LuaUtils.string.endsWith(matchId, 'RxMBR') then
```

Comment:
Would be better to put this in Module:String

Reply:
Problem with that is that Module:String already has a "endswith" function that works differently (with pattern matching and returning "yes"/"" instead of booleans)

Reply:
Will revisit this once we figure out where to put the string methods.